### PR TITLE
Update Xpress interfaces to support 9.5

### LINKED ISSUE #3391: Xpress 9.5 updated Xpress' Python API
## Summary
https://www.fico.com/fico-xpress-optimization/docs/latest/relnotes/GUID-85032F3B-84B8-42A1-A4D4-A0A24FF0A648.html?scroll=GUID-85032F3B-84B8-42A1-A4D4-A0A24FF0A648__SECTION_AWJ_1C5_ZCC

Relevant section:
> Python:
> *  Using deprecated functionality now causes a warning to be emitted. Warnings can be disabled with warnings.simplefilter or warnings.filterwarnings; see the documentation for more information.
> *    New deprecations: creating unlinked variables, sharing unlinked variables between several problems, passing variables or constraints objects to the API while in a callback or presolved, problem.getObjVal, accessing the constraint body via constraint.body.
> *    New function constraint.getBody can be used to get an expression representing the left-hand side of a constraint.
> *    Functions problem.getSlack, problem.getDual and problem.getRCost have been renamed to problem.getSlacks, problem.getDuals and problem.getRedCosts.
> *    The problem.getSolution, problem.getSlacks, problem.getDuals and problem.getRedCosts functions now behave consistently with XPRSgetsolution, XPRSgetslacks, XPRSgetduals and XPRSgetredcosts.
> *    The xpress.Dot operator now supports SciPy matrices of type int32, int64, float32 and float64 in CSR and CSC format.

### Description

We should update the Xpress direct and persistent interfaces to reflect this change. In the meantime maybe we should automatically filter the warnings?

### Additional information

Feel free to assign me this issue.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
Xpress 9.5 was a disruptive release that introduced deprecation warnings and broke a number of tests.  This PR builds on https://github.com/blnicho/pyomo/tree/xpress-9.5.0 and adds a compatibility shim to the underlying `xpress` module to support both pre- and post-9.5 Xpress.

## Changes proposed in this PR:
- Add a compatibility shim to the XpressDirect class for interacting with `xpress`
- Update one test to track a change in the default Xpress behavior
- Update ComponentMap / ComponentSet to expost the internal `hasher` dict and add a `hasher.hashable()` method to query / set the hashability for a given type.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/xpress_direct.py [hunk 1]
```diff
@@ -75,6 +75,100 @@ def _finalize_xpress_import(xpress, avail):
...
+                indices, weights, type, name
+            )
+        )
+        XpressDirect._getSlacks = lambda self, prob, con: prob.getSlacks(con)
+        XpressDirect._getDuals = lambda self, prob, con: prob.getDuals(con)
+        XpressDirect._getRedCosts = lambda self, prob, con: prob.getRedCosts(con)
+
+        # Note that as of 9.5, xpress.var raises an exception when
+        # compared using '==' after it has been removed from the model.
+        # This can fould up ComponentMaps in the persistent interface,
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/xpress_direct.py:167]:
```suggestion
        # This can foul up ComponentMaps in the persistent interface,
```

File: pyomo/solvers/plugins/solvers/xpress_direct.py [hunk 2]
```diff
@@ -894,41 +1005,60 @@ def _postsolve(self):
...
+                    if extract_slacks:
+                        slacks = self._getSlacks(xprob, xpress_cons)
+                except xpress.ModelError:
+                    # Xpress 9.5.0 has new behavior for unbounded
+                    # problems that have mipsols > 0.  Previously
+                    # getSolution() would return a solution, but now
+                    # raises a ModelError (even though the deprecated
+                    # getmipsol() will return a solution).  We will try
+                    # to fall back on the [deprecated] getmipsol(), but
+                    # if it fails, we will raise the original exception.
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/xpress_direct.py:1032]: Is this worth reporting to the Xpress folks? It seems like undesirable behavior, to say the least, and they may eventually remove `getmipsol`.

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/solvers/plugins/solvers/xpress_direct.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/solvers/plugins/solvers/xpress_direct.py:1032]: Yes - it is, once we get our tests passing again.  There were other changes that *looked* like compatibility issues until I got all the deprecation warnings silenced.  (I have some evidence that other deprecation paths are not actually 100% backwards compatible).

**Review by mrmundt — APPROVED** [review]

File: pyomo/solvers/plugins/solvers/xpress_direct.py (hunk 2 above)
Comment (djunglas) [reply pyomo/solvers/plugins/solvers/xpress_direct.py:1032]: Just so that I understand correctly: you are saying that the [MIPSOLS](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/MIPSOLS.html) attribute returns a number greater than zero but still `getSolution()` raises an error?

Comment (jsiirola) [reply pyomo/solvers/plugins/solvers/xpress_direct.py:1032]: That is correct.  It's probably an unintentional regression [and admittedly a pretty weird edge case that we happened to have a test case for] -- filing a bug report upstream is on our backlog (we are still working / fighting infrastructure to try and get a release out this week)

File: pyomo/solvers/tests/checks/test_xpress_persistent.py [hunk 3]
```diff
@@ -291,6 +291,9 @@ def test_nonconvexqp_locally_optimal(self):
 
         opt = pe.SolverFactory('xpress_direct')
         opt.options['XSLP_SOLVER'] = 0
+        # xpress 9.5.0 now defaults to trying (and failing) to solve this problem
+        # using the global solver. This option forces the use of the local solver.
+        opt.options['NLPSOLVER'] = 1
```
Comment (djunglas) [inline pyomo/solvers/tests/checks/test_xpress_persistent.py:296]: This is most likely due to a change in the community license. The community license is the default license you get when you install Xpress via pip or conda and I suppose this is the license you are using for tests.
Pre 9.5 this license did not allow running the Xpress global solver. So Xpress did only aim for finding local optima here.
Starting with 9.5, the community license now allows using the global solver as well (with limits on the problem size). That is why Xpress now uses the global solver by default.

File: pyomo/solvers/plugins/solvers/xpress_direct.py (hunk 2 above)
Comment (djunglas) [reply pyomo/solvers/plugins/solvers/xpress_direct.py:1032]: Thanks. We confirm that this is a defect in the Xpress C library. We will work on it.
